src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -146,90 +191,142 @@ private async ValueTask LoadOrReloadProjectsAsync(ImmutableSegmentedList<Project
 
-            if (!_projectFileExtensionRegistry.TryGetLanguageNameFromProjectPath(projectPath, DiagnosticReportingMode.Ignore, out var languageName))
-                return false;
-
-            var loadedFile = await buildHost.LoadProjectFileAsync(projectPath, languageName, cancellationToken);
-            var diagnosticLogItems = await loadedFile.GetDiagnosticLogItemsAsync(cancellationToken);
+            var diagnosticLogItems = await remoteProjectFile.GetDiagnosticLogItemsAsync(cancellationToken);
             if (diagnosticLogItems.Any(item => item.Kind is DiagnosticLogItemKind.Error))
             {
                 await LogDiagnosticsAsync(diagnosticLogItems);
                 // We have total failures in evaluation, no point in continuing.
                 return false;
             }
 
-            var loadedProjectInfos = await loadedFile.GetProjectFileInfosAsync(cancellationToken);
+            var loadedProjectInfos = await remoteProjectFile.GetProjectFileInfosAsync(cancellationToken);
 
             // The out-of-proc build host supports more languages than we may actually have Workspace binaries for, so ensure we can actually process that
             // language in-process.
             var projectLanguage = loadedProjectInfos.FirstOrDefault()?.Language;
             if (projectLanguage != null && ProjectFactory.Workspace.Services.GetLanguageService<ICommandLineParserService>(projectLanguage) == null)
             {
+                _logger.LogWarning($"Could not run a design-time build for '{projectPath}' because it uses unsupported language '{projectLanguage}'.");
                 return false;
             }
 
-            var existingProjects = _loadedProjects.GetOrAdd(projectPath, static _ => []);
-
             Dictionary<ProjectFileInfo, ProjectLoadTelemetryReporter.TelemetryInfo> telemetryInfos = [];
             var needsRestore = false;
 
-            foreach (var loadedProjectInfo in loadedProjectInfos)
+            using (await _gate.DisposableWaitAsync(cancellationToken))
             {
-                // If we already have the project with this same target framework, just update it
-                var existingProject = existingProjects.Find(p => p.GetTargetFramework() == loadedProjectInfo.TargetFramework);
-                bool targetNeedsRestore;
-                ProjectLoadTelemetryReporter.TelemetryInfo targetTelemetryInfo;
-
-                if (existingProject != null)
+                if (!_loadedProjects.TryGetValue(projectPath, out var currentLoadState))
                 {
-                    (targetTelemetryInfo, targetNeedsRestore) = await existingProject.UpdateWithNewProjectInfoAsync(loadedProjectInfo, _logger);
+                    // Project was unloaded. Do not proceed with reloading it.
+                    return false;
                 }
-                else
+
+                Contract.ThrowIfTrue(currentLoadState is ProjectLoadState.PrimordialOnly);
```

Comment:
this body is a chonker.  can we possibly extact it, make it pure data-in/data-out, and then make any state transition we need at the end?

Reply:
I made a change to simplify the body somewhat, but, I didn't feel there was a practical way to delay updates to shared state until the end. The work which is being done in here is very substantially updating shared state, including the creation of a LoadedProject instance. I am open to suggestions though.